File: cs/Cell.cs
using System; using static System.Numerics.BitOperations; using NUnit.Framework; namespace Sudoku {

readonly struct Cell : IEquatable<Cell> {
    readonly ushort bits_;

    // public constructor
    public Cell( int values ) {
        var bits = 0;
        while( values > 0 ) { bits|= 1 << (values % 10); values/= 10; }
        bits_ = (ushort)(bits >> 1);
    }

    // private constructor
    Cell( ushort bits ) => bits_ = bits;

    // operators
    public static bool operator==( Cell a, Cell b ) => a.bits_ == b.bits_;
    public static bool operator!=( Cell a, Cell b ) => a.bits_ != b.bits_;
    public static Cell operator+( Cell a, Cell b ) => new( (ushort)( a.bits_ | b.bits_ ) ); // union of two cells
    public static Cell operator-( Cell a, Cell b ) => new( (ushort)( a.bits_ & ~b.bits_ ) ); // remove contents of cell 'b' from cell 'a'
    public static Cell operator&( Cell a, Cell b ) => new( (ushort)( a.bits_ & b.bits_ ) ); // intersection of two cells
    public static Cell operator~( Cell a ) => new( (ushort)( ~a.bits_ & 0x1FF ) ); // inverse of a cell

    // properties
    public int Count => PopCount( bits_ );
    public Cell Min => new( (ushort)MinSetBit( bits_ ) );
    public bool Single => IsPow2Or0( bits_ );
    public bool Solvable => 0 != bits_;
    public bool Solved => Single & Solvable;
    
    // converts bit flags back into numbers i.e. 1245
    public int Values { get {
        int value = 0;
        uint bits = bits_;
        for( var place = 1; 0 != bits; place*=10 ) {
            var maxBitIndex = Log2( bits );
            bits&= ~(1U << maxBitIndex); // remove the highest bit
            value+=(maxBitIndex + 1) * place; // add digit to result
        }
        return value;
    }}

    // indexer
    public Cell this[int i] { get {
        uint b = bits_;
        while( i-- > 0 ) b = RemoveMinSetBit( b );
        return new( (ushort)MinSetBit( b ) );
    }}

    // overrides
    public override int GetHashCode() => throw new NotImplementedException( "'cell' type not suitable for hashmap (too many collisions)" );
    public bool Equals( Cell c ) => bits_ == c.bits_;
    public override bool Equals( object o ) => throw new InvalidOperationException( "performance bug (boxing)" );
    public override string ToString() => Values.ToString();

    // utility functions
    static bool IsPow2Or0( int x ) => (x & (~x + 1)) == x;
    static uint MinSetBit( uint x ) => x & ~(x - 1);
    static uint RemoveMinSetBit( uint x ) => x & (x - 1);

    // tests
    [Test] public static void Test() {
        // solved cell
        Cell a = new( 3 );
        Assert.AreEqual( 4, a.bits_ );
        Assert.IsTrue( a.Single );
        Assert.IsTrue( a.Solvable );
        Assert.IsTrue( a.Solved );
        Assert.AreEqual( 1, a.Count );
        Assert.AreEqual( 3, a.Values );
        Assert.AreEqual( a, a.Min );
        Assert.AreEqual( "3", a.ToString() );
        Assert.AreEqual( new Cell( 3 ), a[0] );
        Assert.AreEqual( default( Cell ), a[1] );

        // unsolved cell
        a = new( 125687 );
        Assert.IsFalse( a.Single );
        Assert.IsTrue( a.Solvable );
        Assert.IsFalse( a.Solved );
        Assert.AreEqual( 6, a.Count );
        Assert.AreEqual( new Cell( 1 ), a.Min );
        Assert.AreEqual( 125678, a.Values );
        Assert.AreEqual( new Cell( 1 ), a[0] );
        Assert.AreEqual( new Cell( 2 ), a[1] );
        Assert.AreEqual( new Cell( 5 ), a[2] );
        Assert.AreEqual( new Cell( 6 ), a[3] );
        Assert.AreEqual( new Cell( 7 ), a[4] );
        Assert.AreEqual( new Cell( 8 ), a[5] );
        Assert.AreEqual( default( Cell ), a[6] );
        Assert.AreEqual( "125678", a.ToString() );

        // unsolvable cell
        a = new( 0 );
        Assert.AreEqual( 0, a.bits_ );
        Assert.IsTrue( a.Single );
        Assert.IsFalse( a.Solvable );
        Assert.IsFalse( a.Solved );
        Assert.AreEqual( 0, a.Count );
        Assert.AreEqual( 0, a.Values );
        Assert.AreEqual( default( Cell ), a.Min );
        Assert.AreEqual( default( Cell ), a[0] );
        Assert.AreEqual( "0", a.ToString() );

        // invalid value does not round-trip
        Assert.AreEqual( default( Cell ), new Cell( -1 ) );
        Assert.AreEqual( new Cell( 1 ), new Cell( 10 ) );
        Assert.AreEqual( new Cell( 12 ), new Cell( 121 ) );

        // operators
        Assert.IsTrue( new Cell( 12 ) == new Cell( 21 ) );
        Assert.IsTrue( new Cell( 0 ) != new Cell( 1 ) );
        Assert.AreEqual( new Cell( 1234589 ), new Cell( 12345 ) + new Cell( 98 ) );
        Assert.AreEqual( new Cell( 34 ), new Cell( 12345 ) - new Cell( 1256 ) );
        Assert.AreEqual( new Cell( 7 ), new Cell( 123457 ) & new Cell( 987 ) );
        Assert.AreEqual( new Cell( 13 ), ~new Cell( 2456789 ) );
    }
}

} // namespace

File: rs/src/cell.rs
use std::fmt;
use std::ops;

#[derive( Copy, Clone, Debug, Eq, PartialEq )]
pub struct Cell {
    bits: u16
}

impl Cell {
    // constructors
    pub fn new( mut values: u32 ) -> Cell {
        let mut bits = 0;
        while values > 0 { bits|= 1 << (values % 10); values/=10; }
        Cell { bits: (bits >> 1) as u16 }
    }

    // TODO: can use a constant for this Cell
    pub fn full() -> Cell { Cell { bits: 0x1FF } }
    
    // properties
    pub fn count( &self ) -> u32 { self.bits.count_ones() }
    pub fn min( &self ) -> Cell { Cell { bits: min_bit( self.bits ) } }    
    pub fn single( &self ) -> bool { is_pow2_or_zero( self.bits ) }
    pub fn solvable( &self ) -> bool { self.bits > 0 }    
    pub fn solved( &self ) -> bool { self.single() & self.solvable() }
    
    pub fn values( &self ) -> u32 {
        let mut values = 0;
        let mut place = 1;
        let mut bits = self.bits as u32;
        while bits > 0 {
            let max_bit_index = log2( bits );
            values+=(max_bit_index + 1) * place;
            place*=10;
            bits = clear_bit( bits, max_bit_index );
        }
        values
    }
}

// utility functions
fn clear_bit( x: u32, index: u32 ) -> u32 { x & !(1 << index) }
fn log2( x: u32 ) -> u32 { x.leading_zeros() ^ 31 }
fn is_pow2_or_zero( x: u16 ) -> bool { (x & (!x).wrapping_add( 1 )) == x }
fn min_bit( x: u16 ) -> u16 { x & !x.wrapping_sub( 1 ) }

// serialization
impl fmt::Display for Cell {
    fn fmt( &self, f: &mut fmt::Formatter<'_> ) -> fmt::Result {
        write!( f, "{}", self.values() )
    }
}

// '+' operator gives union of two cells
impl ops::Add for Cell {
    type Output = Cell;
    fn add( self, r: Cell ) -> Cell { Cell { bits: self.bits | r.bits } }
}

impl ops::AddAssign for Cell {
    fn add_assign( &mut self, r: Cell ) { *self = Cell { bits: self.bits | r.bits }; }
}

// '-' operator removes contents of cell
impl ops::Sub for Cell {
    type Output = Cell;
    fn sub( self, r: Cell ) -> Cell { Cell { bits: self.bits & !r.bits } }
}

impl ops::SubAssign for Cell {
    fn sub_assign( &mut self, r: Cell ) { *self = Cell { bits: self.bits & !r.bits } }
}

// '&' operator gives intersection of two cells
impl ops::BitAnd for Cell {
    type Output = Cell;
    fn bitand( self, r: Cell ) -> Cell { Cell { bits: self.bits & r.bits } }
}

impl ops::BitAndAssign for Cell {
    fn bitand_assign( &mut self, r: Cell ) { *self = Cell { bits: self.bits & r.bits } }
}

// '!' operater gives the inverse of a cell
impl ops::Not for Cell {
    type Output = Cell;
    fn not( self ) -> Cell { Cell { bits: !self.bits & 0x1FF } }
}

// tests
#[cfg(test)] mod tests {
    use super::*;

    #[test] fn test() {
        // zero cell
        let c = Cell::new( 0 );
        assert_eq!( 0, c.values() );
        assert_eq!( Cell::new( 0 ), c.min() );
        assert_eq!( 0, c.count() );
        assert!( c.single() );
        assert!( !c.solvable() );
        assert!( !c.solved() );
        assert_eq!( "0", format!( "{}", c ) );

        // solved cell
        let c = Cell::new( 500055550 );
        assert_eq!( 5, c.values() );
        assert_eq!( 1, c.count() );
        assert_eq!( Cell::new( 5 ), c.min() );
        assert!( c.single() );
        assert!( c.solvable() );
        assert!( c.solved() );
        assert_eq!( "5", format!( "{}", c ) );

        // unsolved cell
        let c = Cell::new( 1257007 );
        assert_eq!( 1257, c.values() );
        assert_eq!( 4, c.count() );
        assert_eq!( Cell::new( 1 ), c.min() );
        assert!( !c.single() );
        assert!( c.solvable() );
        assert!( !c.solved() );
        assert_eq!( "1257", format!( "{}", c ) );

        // full cell
        let c = Cell::full();
        assert_eq!( 123456789, c.values() );
        assert_eq!( 9, c.count() );

        // operators
        assert!( Cell::new( 12 ) == Cell::new( 21 ) );
        assert!( Cell::new( 0 ) != Cell::new( 1 ) );
        assert_eq!( Cell::new( 1234589 ), Cell::new( 12345 ) + Cell::new( 98 ) );
        assert_eq!( Cell::new( 34 ), Cell::new( 12345 ) - Cell::new( 1256 ) );
        assert_eq!( Cell::new( 7 ), Cell::new( 123457 ) & Cell::new( 987 ) );
        assert_eq!( Cell::new( 13 ), !Cell::new( 2456789 ) );

        // assignment operators
        let mut c = Cell::new( 123 );
        c+= Cell::new( 4 );
        assert_eq!( Cell::new( 1234 ), c );
        c-= Cell::new( 4 );
        assert_eq!( Cell::new( 123 ), c );
        c&= Cell::new( 13 );
        assert_eq!( Cell::new( 13 ), c );
    }
}
